articles/node-testing/index.md

```diff
@@ -0,0 +1,331 @@
+  }
+
+  console.log("Failed");
+
+}
+
+testAddTwo();
+```
+
+In the example above we used a test case of `x=5` and when we run this script we can see on the console.
+
+```
+Expected 7 Got 7
+Passed
+```
+
+### Structuring Unit Tests
+
+Let's examine one way of structing tests, **AAA Pattern**. A test has 3 parts or sections, Arange, Act, Assert.
+
+1. Arange: Initialize variables, setup necessary preconditions and inputs.
+2. Act: Execute the function/unit being tested.
+3. Assert: *Assert* (or check) that the value received from executing the unit is the same value that is expected.
+
+The code above is
+
+```javascript
+//index.js
+function addTwo(a){
+  return a+2; 	
+}
+
+function testAddTwo() {
+	//Step 1. Arange
+  var x = 5;
+  var x2 = x + 2;
+
+	//Step 2. Act
+  var x3 = addTwo(x);
+
+  console.log("Expected "+x2+" Got " + x3);
+	//Step 3. Assert
+  if (x2 === x3){
+    return console.log("Passed");
+  }
+
+  console.log("Failed");
+
+}
+
+testAddTwo();
+```
+
+The `if` condition in the code example above *asserts* whether the code is working correctly or not. Node.js comes with an [assert module](https://nodejs.org/api/assert.html) with it. Let's do the same action using `assert`.
+
+```javascript
+var assert = require("assert")
+function addTwo(a){
+  return a+2; 	
+}
+function testAddTwo(){
+	var x = 5;
+	var y1 = x + 2;
+	var y2 = addTwo(x);
+	console.log("Expected "+y1+" Got " +y2);
+	try{
+		assert.equal(y1,y2);
+		console.log("Passed");
+	 } catch (err) {
+	 	console.log("Failed");
+		console.log(err);
+		}
+	}
+
+testAddTwo();
+```
+
+We can expect the following result on the console for the correct code.
+
+```
+Expected 7 Got 7
+Passed
+```
+
+When we deliberately change `var y1 = x + 2;` to `var y1 = x + 1;` we see the following error message.
+
+```
+Expected 6 Got 7
+Failed
+AssertionError [ERR_ASSERTION]: 6 == 7
+{
+  generatedMessage: true,
+  code: 'ERR_ASSERTION',
+  actual: 6,
+  expected: 7,
+  operator: '=='
+}
+```
+
+`equal` tests if the contents of a variable are indeed equal to a value specified in the second argument. Let's look at various methods in the assert module and how we can use them.
+
+Examples:
+
+-  `assert.equal(a,b,c)`. Throws error *c* if *a* is not equal to *b*. *c* can be a sentence.
+- `assert.notEqual`. Used when the generation of a certain value by application indicates a problem in logic.
+- `assert.strictEqual`. This uses [strict equality](https://stackoverflow.com/questions/359494/which-equals-operator-vs-should-be-used-in-javascript-comparisons) (===) rather than (==).
+- `assert.deepEqual`. This compares the objects. They recursively compare two objects, comparing two object’s properties and, if the properties are themselves objects, comparing these as well.
+- `assert.ok`. This is used for testing [asynchronous functions](https://bitsofco.de/asynchronous-functions-101/). When we use it as `assert.ok(value, error_message)` if the value is false then it throws the `error_message`.
```

Comment:
> This is used for testing [asynchronous functions]

Not exclusively used for async testing. I would change the wording to "This can be used"